### PR TITLE
Merge feature/logging to master

### PR DESCRIPTION
**Pull request by dbelz**
This is the first implementation of the logging service for CodeMonkeysMVVM. 

@Pantheas We should talk about a few things here. But this version should work. If you find any bugs or similiar while using please contact me.

Daniel

### CODE REVIEW

File: CodeMonkeysMVVM/Interfaces/Services/Logging/ILogEntry.cs [hunk 1]
```diff
@@ -2,17 +2,18 @@
 
 namespace CodeMonkeysMVVM.Interfaces.Services.Logging
 {
-    public interface ILogItemMetadata
+    public interface ILogEntry
     {
+        Guid ServiceGuid { get; }
+
         DateTime Time { get; set; }
```
[the remark this answers is not in the document]
Comment (dbelz) [reply CodeMonkeysMVVM/Interfaces/Services/Logging/ILogEntry.cs]: Resolved

File: CodeMonkeysMVVM/Logging/Core/KnownProviders.cs [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+namespace CodeMonkeysMVVM.Logging.Core
+{
+    internal static class KnownProviders
+    {
+        internal static HashSet<ILogProvider> Instances;
+
+        static KnownProviders()
+        {
+            Instances = new HashSet<ILogProvider>(new ProviderTypeEqualityComparer());
```
[the remark this answers is not in the document]
Comment (dbelz) [reply CodeMonkeysMVVM/Logging/Core/KnownProviders.cs]: Resolved

File: CodeMonkeysMVVM/Logging/Core/LogEngine.cs [hunk 3]
```diff
@@ -0,0 +1,36 @@
+﻿using CodeMonkeysMVVM.Interfaces.Services.Logging;
+using CodeMonkeysMVVM.Logging.Providers;
+using System;
+
+namespace CodeMonkeysMVVM.Logging.Core
+{
+    public static class LogEngine
+    {
+        public const LogLevel DEFAULT_SERVICE_LOG_LEVEL = LogLevel.Debug;
```
[the remark this answers is not in the document]
Comment (dbelz) [reply CodeMonkeysMVVM/Logging/Core/LogEngine.cs]: Resolved

File: CodeMonkeysMVVM/Core/Implementation/Argument.cs [hunk 4]
```diff
@@ -0,0 +1,28 @@
+﻿using System;
+using System.Diagnostics;
+
+namespace CodeMonkeysMVVM.Core.Implementation
```
[the remark this answers is not in the document]
Comment (dbelz) [reply CodeMonkeysMVVM/Core/Implementation/Argument.cs]: No.

File: CodeMonkeysMVVM/Logging/Providers/FileProvider.cs [hunk 5]
```diff
@@ -2,69 +2,62 @@
...
 namespace CodeMonkeysMVVM.Logging.Providers
 {
-    public class FileProvider : ILogProvider
+    public class FileProvider :
+        TargetProviderBase,
+        ILogProvider
     {
-        private readonly DirectoryInfo storageDirectory;
-        private readonly string logFileExtension;
+        private readonly DirectoryInfo _directory;
```
[the remark this answers is not in the document]
Comment (dbelz) [reply CodeMonkeysMVVM/Logging/Providers/FileProvider.cs]: Yes. Resolved

File: CodeMonkeysMVVM/Interfaces/Services/Logging/ILogEntry.cs [hunk 6]
```diff
@@ -2,17 +2,18 @@
 
 namespace CodeMonkeysMVVM.Interfaces.Services.Logging
 {
-    public interface ILogItemMetadata
+    public interface ILogEntry
     {
-        DateTime Time { get; set; }
+        Guid ServiceGuid { get; }
+
+        DateTime TimeStamp { get; set; }
```
Comment (Pantheas) [inline CodeMonkeysMVVM/Interfaces/Services/Logging/ILogEntry.cs]: Do you use UTC-Time?
Please rename to something like "UtcTimeStamp" or similar, if so

File: CodeMonkeysMVVM/Logging/Core/LogEngine.cs [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+
+        public static Type DefaultProviderType { get; }
+            = typeof(TraceProvider);
+
+        public static int DebugEntryBatchSize { get; set; } = 10;
+        public static int InfoEntryBatchSize { get; set; } = 100;
+        public static int WarnEntryBatchSize { get; set; } = 100;
+        public static int ErrorEntryBatchSize { get; set; } = 100;
+
+        internal static int GetBatchSizeForLevel(LogLevel level)
```
Comment (Pantheas) [inline CodeMonkeysMVVM/Logging/Core/LogEngine.cs]: Linebreak please

File: CodeMonkeysMVVM/Logging/Core/LogEntryBuffer.cs [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+                return;
+
+            _entryBuffer[level].Enqueue(entry);
+
+            WhenBatchSizeReachedDequeue(level);
+        }
+
+        private static void WhenBatchSizeReachedDequeue(LogLevel level)
+        {
+            int batchSize = LogEngine.GetBatchSizeForLevel(level);
```
Comment (Pantheas) [inline CodeMonkeysMVVM/Logging/Core/LogEntryBuffer.cs]: imo this should be checked in `EnqueueEntry` and this method should be renamed, as it sounds quite like a unit test.
You could also split this into 2 methods to make naming easier - e.g. `DequeueBatch` and `WriteBatch` or similar

File: CodeMonkeysMVVM/Logging/LogEntry.cs [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+        public LogLevel Level { get; set; }
+
+        public string Message { get; set; }
+        public string MemberName { get; set; }
+        public string FileName { get; set; }
+
+        public int LineNumber { get; set; }
+        public Exception Exception { get; set; }
+
+        public LogEntry(
```
Comment (Pantheas) [inline CodeMonkeysMVVM/Logging/LogEntry.cs]: Does a public constructor make sense?
I don't see a case where LogEntries should be constructed manually, in addition it has quite a few parameters...

**Comment by Pantheas** [comment]
Please merge again with master after my pull request has been approved as it contains some changes regarding project structure

**Pull request merged**